src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -542,7 +542,12 @@ private static ExpressionSyntax SkipParensAndNullSuppressions(ExpressionSyntax e
 
             if (!convertedExpression.HasErrors && !hasErrors)
             {
-                if (constantValueOpt == null)
+                var conversion = convertedExpression as BoundConversion;
+                if (conversion != null && conversion.ConversionKind == ConversionKind.ImplicitUserDefined)
+                {
+                    diagnostics.Add(ErrorCode.ERR_NonConstantConversionInConstantPattern, patternExpression.Location, conversion.Operand.Type!, conversion.Type);
+                }
+                else if (constantValueOpt == null)
                 {
                     diagnostics.Add(ErrorCode.ERR_ConstantExpected, patternExpression.Location);
```

Comment:
>diagnostics.Add(ErrorCode.ERR_ConstantExpected, patternExpression.Location); [](http://example.com/codeflow?start=20&length=76)

Is there really a guarantee that we will never get here for a built-in conversion that starts with a constant? What if instead of singling out user defined conversions, we report an error here that would include the ```inputType```. Something like: "A constant value of type 'XName' is expected". To be clear, I am suggesting this as an alternative to the current change, no special handling for user-defined conversions, etc.

Reply:
Judging by Binder_Patterns.cs lines 667 through 669, a comment describes that user-defined conversions do not have a constant value - so I'd say that part is covered. (please correct me if im wrong). 
Now you mention built-in conversions starting with a constant but implementing a non-constant conversion, I have searched but not found any such conversions, admittedly I am not very familar with the internals of roslyn yet so I'll gladly be corrected on this too. 

If any such conversions may be mentioned, I could write up a test case for it and take it from there.

I did find that the implicit conversion from string to ReadOnlySpan<char> is now also possible within the pattern expression and appears to compile just fine, is this expected behavior or should not be allowed like user-defined conversions?

Besides this I do agree that clarifying the message belonging to ERR_ConstantExpected to include the type of constant expected to be an improvement, perhaps another issue would be suitable for this? I'd gladly pick it up.

Reply:
Reply to https://github.com/dotnet/roslyn/pull/64741#discussion_r996202027

Unresolving the comment thread as I wasn't sure @AlekseyTs saw and had the chance to respond. We will want to agree on an approach before moving forward.

> If [constant-to-non-constant conversions] do not exist I believe the current implementation to be clearer to the user.

I think I agree. A user-defined conversion is a like a method call that occurs implicitly based on how an expression is used. If your pattern is a method call `M(a)`, and the compiler says, "you need a constant value", it seems obvious that the issue is a method call does not have a constant value. If that method call is occurring implicitly because it's a conversion operator, then it's not so obvious. With Aleksey's suggestion you can infer that the conversion from the pattern-expression to the pattern input type is not a constant value, but if you are not already familiar with user-defined conversions in the language, this still might be confusing.

Reply:
Have discussed offline. We think that just including the expected type of the constant value, in this case `inputType`, in the diagnostic message will be sufficient to improve the experience here. This will free us up from needing to maintain or expand on the diagnostic messages for these as the behavior of conversions in the language/compiler evolves over time.

Let's change the new error code to something like `ERR_ConstantValueOfTypeExpected`, and the message to `A constant value of type '{0}' is expected`. Thanks!